keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,218 @@
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Spec holds configuration for the credential provider
+    Spec RegistryCredentialProviderSpec `json:"spec,omitempty"`
+}
+
+type RegistryCredentialProviderSpec struct {
+    // ImageMatchRegex is used to match against the Image URI to determine which images to
+    // provide credentials for
+    ImageMatchRegex string
+
+    // Exec specifies a custom exec-based authentication plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The existing `ExecCredentialSpec` will add support for registry credentials.  In the in-tree implementation, the docker keyring,
+which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.
+
+```go
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+    // +optional
+    Response *Response `json:"response,omitempty"`
+
+    // +optional
+    Interactive bool `json:"interactive,omitempty"`
+
+    // Adding field:
+    // Image is used when passed to registry credential providers to as part of an image pull
+    Image string `json:"interactive,omitempty"`
+}
+
+// ExecCredentialStatus holds credentials for kube runtime to use for image pulls.
+type ExecCredentialStatus struct {
+    // +optional
+    ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+    Token string `json:"token,omitempty"`
+    ClientCertificateData string `json:"clientCertificateData,omitempty"`
+    ClientKeyData string `json:"clientKeyData,omitempty"`
+
+    // Adding field:
+    // AuthConfig information for connecting to a Registry
+    AuthConfig RegistryAuthConfig
+}
+
+// Adding type:
+// RegistryAuthConfig contains authorization information for connecting to a Registry
+// This type mirrors "github.com/docker/docker/api/types.AuthConfig"
+type RegistryAuthConfig struct {
+    Username string `json:"username,omitempty"`
+    Password string `json:"password,omitempty"`
+    Auth     string `json:"auth,omitempty"`
+
+    // Email is an optional value associated with the username.
+    // This field is deprecated and will be removed in a later
+    // version of docker.
+    Email string `json:"email,omitempty"`
+
+    ServerAddress string `json:"serveraddress,omitempty"`
+
+    // IdentityToken is used to authenticate the user and get
+    // an access token for the registry.
+    IdentityToken string `json:"identitytoken,omitempty"`
+
+    // RegistryToken is a bearer token to be sent to a registry
+    RegistryToken string `json:"registrytoken,omitempty"`
+}
+
+```
+
+### Alternatives Considered
```

Comment:
I know we talked about it, but I can't remember the conclusions. Can you add a section discussing whether the Kubelet should provide credentials at all, versus pushing the responsibility down into the image service (CRI)?

Reply:
We didn't come to any specific conclusions during that discussion, but after discussing it in sig-cloudprovider, we didn't see any benefit there that outweighed the fact that it seemed like a more significant redesign and undertaking, and we preferred to have something that could go alpha in 1.18.